// ircv3/lib/Message/MessageTypes/Commands/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// region RFC1459
// region 4 Message details
// region 4.1 Connection Registration
// 4.1.1 Password message
var Password_1 = require("./Password");
exports.Password = Password_1.default;
// 4.1.2 Nickname message
var NickChange_1 = require("./NickChange");
exports.NickChange = NickChange_1.default;
// 4.1.3 User message
var UserRegistration_1 = require("./UserRegistration");
exports.UserRegistration = UserRegistration_1.default;
// 4.1.4 Server message
// We can't be a server yet
// 4.1.5 Operator message
var OperLogin_1 = require("./OperLogin");
exports.OperLogin = OperLogin_1.default;
// 4.1.6 Quit message
var ClientQuit_1 = require("./ClientQuit");
exports.ClientQuit = ClientQuit_1.default;
// 4.1.7 Server Quit message
var ServerQuit_1 = require("./ServerQuit");
exports.ServerQuit = ServerQuit_1.default;
// endregion
// region 4.2 Channel operations
// 4.2.1 Join message
var ChannelJoin_1 = require("./ChannelJoin");
exports.ChannelJoin = ChannelJoin_1.default;
// 4.2.2 Part message
var ChannelPart_1 = require("./ChannelPart");
exports.ChannelPart = ChannelPart_1.default;
// 4.2.3 Mode message
var Mode_1 = require("./Mode");
exports.Mode = Mode_1.default;
// 4.2.4 Topic message
var Topic_1 = require("./Topic");
exports.Topic = Topic_1.default;
// 4.2.5 Names message
var Names_1 = require("./Names");
exports.Names = Names_1.default;
// 4.2.6 List message
var ChannelList_1 = require("./ChannelList");
exports.ChannelList = ChannelList_1.default;
// 4.2.7 Invite message
var ChannelInvite_1 = require("./ChannelInvite");
exports.ChannelInvite = ChannelInvite_1.default;
// 4.2.8 Kick message
var ChannelKick_1 = require("./ChannelKick");
exports.ChannelKick = ChannelKick_1.default;
// endregion
// region 4.3 Server queries and commands
// TODO
// endregion
// region 4.4 Sending messages
// 4.4.1 Private messages
var PrivateMessage_1 = require("./PrivateMessage");
exports.PrivateMessage = PrivateMessage_1.default;
// 4.4.2 Notice messages
var Notice_1 = require("./Notice");
exports.Notice = Notice_1.default;
// endregion
// region 4.5 User-based queries
// 4.5.1 Who query
var WhoQuery_1 = require("./WhoQuery");
exports.WhoQuery = WhoQuery_1.default;
// 4.5.2 Whois query
var WhoIsQuery_1 = require("./WhoIsQuery");
exports.WhoIsQuery = WhoIsQuery_1.default;
// 4.5.3 Whowas
var WhoWasQuery_1 = require("./WhoWasQuery");
exports.WhoWasQuery = WhoWasQuery_1.default;
// endregion
// region 4.6 Miscellaneous messages
// 4.6.1 Kill message
var Kill_1 = require("./Kill");
exports.Kill = Kill_1.default;
// 4.6.2 Ping message
var Ping_1 = require("./Ping");
exports.Ping = Ping_1.default;
// 4.6.3 Pong message
var Pong_1 = require("./Pong");
exports.Pong = Pong_1.default;
// 4.6.4 Error message
var Error_1 = require("./Error");
exports.Error = Error_1.default;
// endregion
// endregion
// region 5 Optionals
// 5.1 Away
var Away_1 = require("./Away");
exports.Away = Away_1.default;
// 5.2 Rehash message
var Rehash_1 = require("./Rehash");
exports.Rehash = Rehash_1.default;
// 5.3 Restart message
var Restart_1 = require("./Restart");
exports.Restart = Restart_1.default;
// 5.4 Summon message
// really, would anyone use that?
// 5.5 Users message
// would anyone WANT to have that? it seems like a huge security risk.
// 5.6 Operwall message
var WallopsMessage_1 = require("./WallopsMessage");
exports.Wallops = WallopsMessage_1.default;
// 5.7 Userhost message
var UserHostQuery_1 = require("./UserHostQuery");
exports.UserHostQuery = UserHostQuery_1.default;
// 5.8 Ison message
var IsOnQuery_1 = require("./IsOnQuery");
exports.IsOnQuery = IsOnQuery_1.default;
// endregion
// endregion
// region IRCv3
// Capability Negotiation
var CapabilityNegotiation_1 = require("./CapabilityNegotiation");
exports.CapabilityNegotiation = CapabilityNegotiation_1.default;
// endregion
