// ircv3/lib/Message/MessageTypes/Numerics/Error421UnknownCommand.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Message_1 = require("../../Message");
var Error421UnknownCommand = /** @class */ (function (_super) {
    __extends(Error421UnknownCommand, _super);
    function Error421UnknownCommand() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Error421UnknownCommand.prototype.isResponseTo = function (originalMessage) {
        return originalMessage.command === this.params.command;
    };
    Error421UnknownCommand.prototype.endsResponseTo = function (originalMessage) {
        return true;
    };
    Error421UnknownCommand.COMMAND = '421';
    Error421UnknownCommand.PARAM_SPEC = {
        me: {},
        command: {},
        suffix: {
            trailing: true
        }
    };
    return Error421UnknownCommand;
}(Message_1.default));
exports.default = Error421UnknownCommand;
