// ircv3/lib/Message/MessageTypes/Numerics/Reply004ServerInfo.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Message_1 = require("../../Message");
var Reply004ServerInfo = /** @class */ (function (_super) {
    __extends(Reply004ServerInfo, _super);
    function Reply004ServerInfo() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Reply004ServerInfo.COMMAND = '004';
    Reply004ServerInfo.PARAM_SPEC = {
        me: {},
        serverName: {},
        version: {
            optional: true
        },
        userModes: {
            optional: true
        },
        channelModes: {
            optional: true
        },
        channelModesWithParam: {
            optional: true
        }
    };
    return Reply004ServerInfo;
}(Message_1.default));
exports.default = Reply004ServerInfo;
