// ircv3/lib/Message/MessageTypes/Numerics/Reply353NamesReply.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Message_1 = require("../../Message");
var Commands_1 = require("../Commands");
var Reply353NamesReply = /** @class */ (function (_super) {
    __extends(Reply353NamesReply, _super);
    function Reply353NamesReply() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Reply353NamesReply.prototype.isResponseTo = function (originalMessage) {
        return originalMessage instanceof Commands_1.Names;
    };
    Reply353NamesReply.COMMAND = '353';
    Reply353NamesReply.PARAM_SPEC = {
        me: {},
        sentinel: {},
        channel: {
            type: 'channel'
        },
        names: {
            trailing: true
        }
    };
    return Reply353NamesReply;
}(Message_1.default));
exports.default = Reply353NamesReply;
