// ircv3/lib/Message/MessageTypes/Numerics/Reply366EndOfNames.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Message_1 = require("../../Message");
var Commands_1 = require("../Commands");
var Reply366EndOfNames = /** @class */ (function (_super) {
    __extends(Reply366EndOfNames, _super);
    function Reply366EndOfNames() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Reply366EndOfNames.prototype.isResponseTo = function (originalMessage) {
        return originalMessage instanceof Commands_1.Names;
    };
    Reply366EndOfNames.prototype.endsResponseTo = function (originalMessage) {
        return true;
    };
    Reply366EndOfNames.COMMAND = '366';
    Reply366EndOfNames.PARAM_SPEC = {
        me: {},
        channel: {
            type: 'channel'
        },
        suffix: {
            trailing: true
        }
    };
    return Reply366EndOfNames;
}(Message_1.default));
exports.default = Reply366EndOfNames;
